src/coreclr/src/vm/nativeimage.cpp

```diff
@@ -0,0 +1,156 @@
+// Headers
+// --------------------------------------------------------------------------------
+
+#include <shlwapi.h>
+
+BOOL AssemblyNameIndexHashTraits::Equals(LPCUTF8 a, LPCUTF8 b)
+{
+    WRAPPER_NO_CONTRACT;
+
+    SString s1;
+    SString s2;
+    s1.SetUTF8(a);
+    s2.SetUTF8(b);
+    return s1.CompareCaseInsensitive(s2) == 0;
+}
+
+AssemblyNameIndexHashTraits::count_t AssemblyNameIndexHashTraits::Hash(LPCUTF8 s)
+{
+    WRAPPER_NO_CONTRACT;
+
+    SString s1;
+    s1.SetUTF8(s);
+    return s1.HashCaseInsensitive();
+}
+
+NativeImage::NativeImage(
+    NewHolder<PEImageLayout>& peImageLayout,
+    READYTORUN_HEADER *pHeader,
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator,
+    AllocMemTracker *pamTracker)
+    : m_eagerFixupsLock(CrstNativeImageEagerFixups)
+{
+    CONTRACTL
+    {
+        THROWS;
+        CONSTRUCTOR_CHECK;
+        STANDARD_VM_CHECK;
+        INJECT_FAULT(COMPlusThrowOM(););
+    }
+    CONTRACTL_END;
+
+    LoaderHeap *pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+    m_utf8SimpleName = nativeImageName;
+    m_peImageLayout.Assign(peImageLayout.Extract());
+    m_eagerFixupsHaveRun = false;
+
+    m_pReadyToRunInfo.Assign(new ReadyToRunInfo(/*pModule*/ NULL, peImageLayout, pHeader, /*compositeImage*/ NULL, pamTracker), /*takeOwnership*/ TRUE);
+    m_pComponentAssemblies = m_pReadyToRunInfo->FindSection(ReadyToRunSectionType::ComponentAssemblies);
+    m_componentAssemblyCount = m_pComponentAssemblies->Size / sizeof(READYTORUN_COMPONENT_ASSEMBLIES_ENTRY);
+    
+    // Check if the current module's image has native manifest metadata, otherwise the current->GetNativeAssemblyImport() asserts.
+    m_pManifestMetadata = LoadManifestMetadata();
+
+    HENUMInternal assemblyEnum;
+    HRESULT hr = m_pManifestMetadata->EnumAllInit(mdtAssemblyRef, &assemblyEnum);
+    mdAssemblyRef assemblyRef;
+    int assemblyIndex = 0;
+    while (m_pManifestMetadata->EnumNext(&assemblyEnum, &assemblyRef))
+    {
+        LPCSTR assemblyName;
+        hr = m_pManifestMetadata->GetAssemblyRefProps(assemblyRef, NULL, NULL, &assemblyName, NULL, NULL, NULL, NULL);
+        m_assemblySimpleNameToIndexMap.Add(AssemblyNameIndex(assemblyName, assemblyIndex));
+        assemblyIndex++;
+    }
+}
+
+NativeImage::~NativeImage()
+{
+    if (m_pManifestMetadata != NULL)
+    {
+        m_pManifestMetadata->Release();
+    }
+}
+
+bool NativeImage::Matches(LPCUTF8 utf8SimpleName) const
+{
+    return !_stricmp(utf8SimpleName, m_utf8SimpleName);
+}
+
+#ifndef DACCESS_COMPILE
+NativeImage *NativeImage::Open(
+    LPCWSTR fullPath,
+    LPCUTF8 nativeImageName,
+    LoaderAllocator *pLoaderAllocator,
+    AllocMemTracker *pamTracker)
+{
+    NewHolder<PEImageLayout> peLoadedImage = PEImageLayout::LoadNative(fullPath);
+
+    uint32_t headerRVA = peLoadedImage->GetExport("RTR_HEADER");
+    if (headerRVA + sizeof(READYTORUN_HEADER) > peLoadedImage->GetSize())
```

Comment:
This should not be a silent failure. We should throw exception here.

Reply:
Instead of sizes, it would be better to check R2R Signature and Version

Reply:
Is there a possibility that the `headerRVA` value, from a corrupted/maliciously-crafted binary, could be overflown when adding it to `sizeof(READYTORUN_HEADER)` in such a way that this check is bypassed?

Usually, rewriting such checks prioritizing compile-time constants and doing subtractions instead of additions will often suffice:

```C
    if (peLoadedImage->GetSize() - sizeof(READYTORUN_HEADER) <= headerRVA)
    {
          ...
    }
```

Reply:
If you get corrupted or maliciously crafted binary here, you can cause arbitrary or anything else.

This should focus on high-value sanity checks that filter out user mistakes. Checking the R2R signature and Version is example of high-value sanity check.

Reply:
Added signature and version check throwing a failure on mismatch in 16th commit.